File: php/mergeSort.php
<?php
// マージソート
function margeSort(&$arr) {
  function m (&$arr, $head, $tail) {
    if ($head < $tail) {
      $mid = floor(($head + $tail) / 2);
      m($arr, $head, $mid);
      m($arr, $mid+1, $tail);
  
      $l = array_slice($arr, $head, $mid+1 - $head);
      $r = array_slice($arr, $mid+1, ($tail - $mid));
      $i = $head;
      while (count($l) > 0 && count($r) > 0) {
        if ($l[0] <= $r[0]) {
          $arr[$i++] = array_shift($l);
        } else {
          $arr[$i++] = array_shift($r);
        }
      }
      while (count($l) > 0) {
        $arr[$i++] = array_shift($l);
      }
    }
    return $arr;
  }
  return m($arr, 0, count($arr)-1);
}

$arr = array(651, 148 , 124, 638, 567, 435, 185, 413, 841, 35);
print_r(margeSort($arr));

File: swift/mergeSort.swift
// マージソート
import Foundation
import Glibc

func margeSort(a: [Int]) -> [Int] {
    var arr = a
    func m(head: Int, tail: Int) {
        if head < tail {
            let mid = Int(floor(Double(head + tail) / 2))
            m(head: head, tail: mid)
            m(head: mid + 1, tail: tail)
    
            var l = arr[head...mid] // インデックスは0番からではなく、headの値からとなる
            var i = head
            var j = mid + 1
            var k = head
            while i <= mid && j <= tail {
                if l[i] <= arr[j] {
                    arr[k] = l[i]
                    i += 1
                } else {
                    arr[k] = arr[j]
                    j += 1
                }
                k += 1
            }
    
            while i <= mid {
                arr[k] = l[i]
                i += 1
                k += 1
            }
        }
    }
    m(head: 0, tail: arr.count-1)
    return arr
}
print(margeSort(a: [867, 104, 413, 541, 304, 354, 165, 150]))
